src/pdu_loop/frame_element.rs

```diff
@@ -0,0 +1,494 @@
+use crate::{
+    command::Command,
+    error::{Error, PduError},
+    pdu_loop::{frame_header::FrameHeader, pdu_flags::PduFlags},
+    ETHERCAT_ETHERTYPE, MASTER_ADDR,
+};
+use cookie_factory::{
+    bytes::{le_u16, le_u8},
+    combinator::{skip, slice},
+    gen_simple, GenError,
+};
+use core::{
+    future::Future,
+    marker::PhantomData,
+    mem,
+    ops::Deref,
+    ptr::{addr_of, addr_of_mut, NonNull},
+    sync::atomic::Ordering,
+    task::Poll,
+};
+use packed_struct::PackedStruct;
+use smoltcp::wire::{EthernetAddress, EthernetFrame};
+use std::task::Waker;
+
+use super::PduResponse;
+
+#[atomic_enum::atomic_enum]
+#[derive(PartialEq, Default)]
+pub enum FrameState {
+    // SAFETY: Because we create a bunch of `Frame`s with `MaybeUninit::zeroed`, the `None` state
+    // MUST be equal to zero. All other fields in `Frame` are overridden in `replace`, so there
+    // should be no UB there.
+    /// The frame is ready to be claimed
+    #[default]
+    None = 0,
+    /// The frame is claimed and can be initialised ready for sending.
+    Created = 1,
+    /// The frame is ready to send when the TX loop next runs.
+    Sendable = 2,
+    /// The frame is being sent over the network interface.
+    Sending = 3,
+    /// A frame response has been received and is now ready for parsing.
+    RxBusy = 5,
+    /// Frame response parsing is complete. The frame and its data is ready to be returned in
+    /// `Poll::Ready`.
+    RxDone = 6,
+    /// The frame TX/RX is complete, but the frame is still in use by calling code.
+    RxProcessing = 7,
+}
```

Comment:
Fewer variants than in https://github.com/jamesmunns/ethercrab/compare/master...jamesmunns:ethercrab:james-hacking as I wasn't sure I needed the granularity. `WAIT_RX` and `DONE` are removed.